servant/src/Servant/Types/SourceT.hs

```diff
@@ -305,13 +306,28 @@ fromAction stop action = SourceT ($ fromActionStep stop action)
 {-# INLINE fromAction #-}
 
 fromActionStep :: Functor m => (a -> Bool) -> m a -> StepT m a
-fromActionStep stop action = loop where
-    loop = Effect $ fmap step action
-    step x
-        | stop x    = Stop
-        | otherwise = Yield x loop
+fromActionStep stop action = fromConsumableActionStep action' () where
+  action' = const $ flip fmap action $
+    \x -> if stop x
+          then Nothing
+          else Just (x,())
 {-# INLINE fromActionStep #-}
 
+-- | Create a `StepT' from a consumable @c@, that is to say from an input and an action that returns
+-- and is called again on an updated version of that input.
```

Comment:
Could you perhaps include a tiny teeny example in the haddocks here? That just unfolds a stream of 10 numbers or something.

Reply:
It seems that the `doctest` suite passes even if I write faulty examples, so I'm not sure. But I checked in a `ghci` and this should be ok.